ls1intum/Artemis
# `Programming exercises`: Configure GitLab for supporting GitLab CI

### Checklist
#### General


#### Server

#### Changes affecting Programming Exercises

### Motivation and Context
As part of my thesis, I add the support of GitLab CI as a CI system for Artemis.
This is the first PR for configuring the CI correctly, when creating and starting an exericse in Artemis.

### Description
I changed just the minimum for adding support for GitLab CI. I implemented the necessary interfaces required for creating and starting the exercises.

### Steps for Testing
Prerequisites:
- Local GitLab instance with a Runner (See #4881 for setup)
- MySQL database

**Please test this pull request locally.**

1. Generate a personal access token with all scopes under http://localhost/-/profile/personal_access_tokens
2. Create `application-local.yml` and adapt the password and the access token
```
artemis:
    course-archives-path: ./exports/courses
    repo-clone-path: ./repos
    repo-download-clone-path: ./repos-download
    encryption-password: artemis_admin           # LEGACY: arbitrary password for encrypting database values
    bcrypt-salt-rounds: 11  # The number of salt rounds for the bcrypt password hashing. Lower numbers make it faster but more unsecure and vice versa.
    # Please use the bcrypt benchmark tool to determine the best number of rounds for your system. https://github.com/ls1intum/bcrypt-Benchmark
    user-management:
        use-external: false
        internal-admin:
            username: artemis_admin
            password: artemis_admin
        accept-terms: false
        login:
            account-name: TUM
    version-control:
        url: http://172.17.0.1:80
        user: root
        password: artemis_admin # Change here!
        token: artemis-gitlab-token # Change here!
    git:
        name: Artemis
        email: artemis.in@tum.de
server:
    port: 8080
    url: http://172.17.0.1:8080 # `http://host.docker.internal:8080` for Windows

```

3. Start Artemis with the `gitlabci` profile. Use `Artemis (Server, GitLabCI & Gitlab)` in IntelliJ.
4. Open Artemis in your browser and create a new course.
5. Generate a new programming exericise (with default configuration).
6. Open the template, solution and test repository. In the template and solution repository CI should be enabled and configured. Go to `Settings` > `CI/CD` > `General Pipelines` and check if `https://home.in.tum.de/~scbe/Artemis/gitlabci-runners/getting-started-template.yml` is the `CI/CD configuration file`. In the test repository CI should be disabled. Check if in `Settings` > `General` > `Visibility, project features, permissions` `CI/CD` is disabled.

### Review Progress


#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| GitLabCIUserManagementService.java | - | 25% |
| GitLabCIService.java  | - | 94% |
| GitLabCIAuthorizationInterceptor.java  | - | 30% |
| GitLabCIBuildPlanLinkInfoContributor.java  | - | 30% |
| GitLabCIProgrammingLanguageFeatureService.java  | - | 100% |
| GitLabCIException | - | 30% |
